config/daemon/ipv6.md

```diff
@@ -10,31 +10,110 @@ Before you can use IPv6 in Docker containers or swarm services, you need to
 enable IPv6 support in the Docker daemon. Afterward, you can choose to use
 either IPv4 or IPv6 (or both) with any container, service, or network.
 
-> **Note**: IPv6 networking is only supported on Docker daemons running on Linux
-> hosts.
+> **Note**
+>
+> - IPv6 support is experimental, use it with caution.
+> - IPv6 is only supported on Docker daemons running on Linux hosts.
 
-1.  Edit `/etc/docker/daemon.json`, set the `ipv6` key to `true` and the `fixed-cidr-v6` key
-    to your IPv6 subnet. In this example we are setting it to `2001:db8:1::/64`.
+## Daemon configuration
 
-    ```json
-    {
-      "ipv6": true,
-      "fixed-cidr-v6": "2001:db8:1::/64"
-    }
-    ```
+To enable IPv6, you must edit the Docker daemon configuration file located at
+`/etc/docker/daemon.json`. Configure the following parameters:
 
-    Save the file.
+```json
+{
+  "experimental": true,
+  "ipv6": true,
+  "ip6tables": true,
+  "fixed-cidr-v6": "2001:db8:1::/64",
+  ...
+}
+```
 
-2.  Reload the Docker configuration file.
+This configuration makes IPv6 networking function as you would expect it to.
+The `ipv6` and `fixed-cidr-v6` parameters are optional.
+They assign an IPv6 subnet to the default bridge network.
 
-    ```console
-    $ systemctl reload docker
-    ```
+After saving the configuration file, restart the Docker daemon for your
+changes to take effect:
 
-You can now create networks with the `--ipv6` flag and assign containers IPv6
-addresses using the `--ip6` flag.
+```console
+$ systemctl restart docker
+```
+
+Upon restart, the daemon assigns IPv6 addresses to containers connected to the
+default bridge network, and to user-defined networks configured with an IPv6 subnet.
```

Comment:
Users need to set `default-address-pools` to let the daemon assign IPv6 addresses to their custom network (as described below).

```suggestion
default bridge network.
```

Reply:
> Users need to set `default-address-pools` to let the daemon assign IPv6 addresses

`docker network create --ipv6 --subnet fd00:d0ca::/112`?

You can assign an IPv6 subnet explicitly for user-defined networks without `default-address-pools` being adjusted.

The statement doesn't say anything about automatically configuring IPv6 subnet for a user-defined network, just that containers with a user-defined network that has already been configured with an IPv6 subnet would enable the container to be assigned an IPv6 address from it.

Reply:
> It talks about automatically assigning IPv6 addresses to containers:

Which is correct. That happens when the network (default bridge, or user-defined nework) has IPv6 subnet already configured?

> My point is: `experimental` and `ip6tables` are needed to have the same experience with IPv6 as with IPv4 (wrt. isolation & port mapping), whether users rely on static or dyanmic allocation.

Yes these two settings are quite important. It's unfortunate `ip6tables` remains opt-in while `userland-proxy: true` is the default due to surprises that causes on an IPv6 host even when docker networks are IPv4 only.

>  Above options won't turn on dynamic allocation for custom networks, hence my suggestion.

I tried to communicate this distinction earlier in review, the two settings from the earlier JSON snippet could belong in their own section that enables IPv6 for the default (legacy) bridge.

A section that follows after it could then cover the same for user-defined networks with an explicit subnet given with `docker network create` or in the `compose.yaml`. After that, you detail implicit subnet assignment which almost works like creating IPv4 networks, but still requires opt-in for IPv6 (`--ipv6` / `enable_ipv6: true`).

> I also prefer to just tell users to configure `ipv6` and `fixed-cidr-v6` such that the default bridge and custom networks just work the same way.

I understand :)

I just think it's worth the distinction that the two settings are optional, only related to enabling the support on the bridge. Some users aren't aware of that in the past and think it's required even when using `compose.yaml`.

Reply:
> Which is correct. That happens when the network (default bridge, or user-defined nework) has IPv6 subnet already configured?

Right, but that's unrelated to the daemon parameters we're advertising in this section.

> I tried to communicate this distinction earlier in review, the two settings from the earlier JSON snippet could belong in their own section that enables IPv6 for the default (legacy) bridge.

Got you. @dvdksn WDYT about what @polarathene proposes:

* Adding a section dedicated to the default bridge and moving `ipv6` and `fixed-cidr-v6` there ;
* Adding a section dedicated to static allocation ;

We can discuss this proposal on Slack if you're unsure what should go in these new sections.

Reply:
Great, I think this would help clear things up. I'll give it it a go!

Reply:
ptal 2be681dea9a6c3689aa61a2ab5cf9b1f2553dad9